### ISSUE TITLE
Why TestCase with "\uffff" argument doesn't work?

### ISSUE THREAD

**Issue by Denis535** [nunit/nunit3-vs-adapter#761]
[TestCase( "\uffff" )]
    public void Test_01(string value) {
        TestContext.WriteLine( value );
    }

After running this test stays white. Neither green nor red, just white. Why? It is very strange.

**Comment by mikkelbu** [comment]
@Denis535 I've transfered the issue to the adapter repo

**Comment by OsirisTerje** [comment]
@Denis535 @mikkelbu  Yes, this is part of the FQN issues, of which we have multiple.  We're currently blocked for most of these issues.  We can fix a few, but not all.  If you have tried 3.16.0 and it doesn't work, and also 3.16.1 doesn't work, you're in that group. 
I'll add this one to the growing list of FQN issues.

**Comment by ChrisMaddock** [comment]
I think this is a duplicate of https://github.com/nunit/nunit/issues/3063#issuecomment-432788447 - would you agree?

**Comment by OsirisTerje** [comment]
Might be yes...  

I checked this with the NUnit.Console now, and that one finds ALL test cases here:

![image](https://user-images.githubusercontent.com/203432/85058255-1dbd8780-b1a2-11ea-83c6-dfd517c48e15.png)

So then I am a bit stumped.  Not sure why the adapter gets a different result than the console.

**Comment by ChrisMaddock** [comment]
Oh, that's odd.

Can you with the console? (Using whatever filter the adapter would normally do)

**Comment by OsirisTerje** [comment]
I just run 
```
nunit3-console fqn.dll
```
No filter

**Comment by ChrisMaddock** [comment]
So, if I'm reading the adapter code right, the adapter's dump is built up by combining all the received test events?

That's a different from the console TestResult.xml, which is passed back from the framework in one piece. That suggests there might be an issue with the TestEvent writing functionality in the framework, which isn't present in the xml writing for the final result.

**Comment by OsirisTerje** [comment]
Not quite.  We first do a discovery during the execution, and there we get a complete list of the testcases, just the same as the the first part of the testresult.xml.  Then we run the execution, which builds it up using the events

**Comment by OsirisTerje** [comment]
[E_FQN.dll.zip](https://github.com/nunit/nunit3-vs-adapter/files/4800572/E_FQN.dll.zip)

This is the complete dump file.  You see both phases here, there are markers between them

**Comment by ChrisMaddock** [comment]
Ah ok. So if you run `nunit3-console fqn.dll --explore`, do you get the test listed?

(Sorry - there's a limited amount I can test myself on this tablet!)

**Comment by OsirisTerje** [comment]
![image](https://user-images.githubusercontent.com/203432/85060508-a853b600-b1a5-11ea-9ac6-409fff776d5e.png)

Yes

**Comment by ChrisMaddock** [comment]
I'm a little suspicious of AsString - only because it's the only notable difference I can see anywhere! Are you able to breakpoint in the debugger, and examine `discoveryResults` before it goes into AsString()?

**Comment by OsirisTerje** [comment]
Yep, did that now, and this output is directly from the runner.Explore
```xml
<test-suite type="ParameterizedMethod" id="0-1024" name="Test_01" fullname="FQN.Issue761.Test_01" classname="FQN.Issue761" runstate="Runnable" testcasecount="2"><test-case id="0-1010" name="Test_01(&quot;￰&quot;)" fullname="FQN.Issue761.Test_01(&quot;￰&quot;)" methodname="Test_01" classname="FQN.Issue761" runstate="Runnable" seed="1850039790" /><test-case id="0-1012" name="Test_01(&quot;ᄑ&quot;)" fullname="FQN.Issue761.Test_01(&quot;ᄑ&quot;)" methodname="Test_01" classname="FQN.Issue761" runstate="Runnable" seed="340842816" /></test-suite>
```

**Comment by OsirisTerje** [comment]
@ChrisMaddock You did mention some time back something about using the api and not the engine directly, that there are some differences.  I don't remember if we did something about that or not.  We had issues, which I also don't exactly remember about the ability to use extensions.  I think that is still open.  Can there be something there about the results we get abck, that we're using the wrong interface somehow ?

**Comment by ChrisMaddock** [comment]
That was my go-to thought to be honest, but from looking at how the adapter does use the engine, I can't think of anything that would be a problem around this... 

I'll see if I can get a chance to have a look at this over the weekend.

**Comment by ChrisMaddock** [comment]
I've ran the debugger, I think this is a duplicate of nunit/nunit#3063.

I'm not too sure what you were seeing last night, Terje - when I run the tests, the \ufff test is discovered fine. It's only on execution that we don't get a result, which is consistent with the above issue and the fact that \ufff can't be constructed into a valid filter for a test name at the moment. @jnm2 explains the problems nicely on that issue.

**Comment by ChrisMaddock** [comment]
FWIW, you get the equivalent effect if you try and filter on the test in the Console.

```
"C:\Users\Chris\Documents\git\nunit3-vs-adapter.issues\Issue640\bin\Debug\net472\FQN.dll" --test FQN.Issue761.Test_01("\uffff")
```

**Comment by OsirisTerje** [comment]
Thanks @ChrisMaddock .  @jnm2  suggest filtering these in the adapter.  Can do that, but as long as the nunit issue remains, it will still be broken. Changing the name in the adapter never works well.  
.----  and,  The test is skipped during execution.

**Comment by dannypike** [comment]
I'm running NUnit Adapter 3.17.0.0, on "Visual Studio 16.7.0 Preview 3.1" and I get the same problem with "\u0001", "\u0010" and "\u0083" but it works OK with "\u0007".

Wondering out loud: might it it have anything specific to do with Unicode-only control characters - i.e. `char.IsControl()` returns **true** - but where they do not have equivalent ASCII escape code sequences?

For example, the value `"\u0007"` above can also be written as `"\a"`, and codes like `"\u000d"` and `"\u000a"` also map directly to ASCII codes `"\r"` and `"\n"`. All of those work, but the ones that have no dedicated ASCII escape codes seem to fail.

**Comment by OsirisTerje** [comment]
Thanks @dannypike , interesting observation.

**Comment by felipepessoto** [comment]
I have the same issue. Unicode chars are pretty common test case argument in unit tests, I think this is high pri since tests are just ignored and you can easily miss it from CI pipeline.

        [TestCase("\u0006")]
        [TestCase("\u009A")]
        [TestCase("\u009B")]
        [TestCase("\u009C")]
        [TestCase("\u009D")]
        [TestCase("\u009E")]
        [TestCase("\u009F")]
        [TestCase("\u009C")]
        [TestCase("OK")]
        public void TestUnicode(string value)
        {
            if(value == "\u009B")
            {
                throw new System.Exception("ABC");
            }
            Assert.Pass();
        }


Maybe the problem is in nunit console, because running

    nunit3-console.exe "TestProject1.dll" --test TestProject1.Tests.TestUnicode(\"\u009b\")

Show it correctly receive the test name, but find 0 tests:

![image](https://user-images.githubusercontent.com/1336227/135737959-3534d45d-5bc5-4882-a0ab-80d036ac75bc.png)

![image](https://user-images.githubusercontent.com/1336227/135737962-f776df4d-e3de-4068-b0b3-941406831dfd.png)

**Comment by felipepessoto** [comment]
Kind of found the problem, the double slash:

![image](https://user-images.githubusercontent.com/1336227/135741899-6e04552f-8627-4e4f-9298-28e2572fdcb4.png)

If I debug and remove one of them (NUnit3FrameworkDriver.CreateObject), it works:

![image](https://user-images.githubusercontent.com/1336227/135741936-c7683dda-d2ef-434e-aacb-977bada87bff.png)

**Comment by stevenaw** [comment]
The `\uffff` scenario mentioned in the title also appears to be discussed a bit on the framework side in https://github.com/nunit/nunit/issues/4826
